csrc/opaque_type.h

```diff
@@ -79,7 +82,11 @@ class Opaque {
 
 template <typename T>
 bool OpaqueEquals<T>::operator()(const Opaque& a, const Opaque& b) const {
-  return a.as<T>() == b.as<T>();
+  if constexpr (dynamic_type::opcheck<T> == dynamic_type::opcheck<T>) {
+    return a.as<T>() == b.as<T>();
+  } else {
+    return std::memcmp(&a.as<T>(), &b.as<T>(), sizeof(T)) == 0;
```

Comment:
You are absolutely right that this implementation will not always correctly compare objects, and your example totally make senses. The problem here is, because C++ is a statically typed language, there is no general way to compare opaque values. So we can only have an approximation of `operator==`, which does not always work. For example, `1 == 1.0`, but `Opaque(1) != Opaque(1.0)`.  This PR adds another approximation, and just like any other approximation, it is not always correct. The only guarantee we can have is: if you are comparing two opaque objects with identical underlying type, and that underlying type defined `operator==`, then you will get the desired behavior.

The problem of bitwise comparison is not just in the example you mentioned. For example, if I have
```C++
struct A {
  int32_t i;
  double d;
};
```
Then usually C++ compiler will allocate memory like this
```
[======== i (32bits) ========][======== empty (32bits) ========][======== d (64bits) ========]
```
And the empty bits will be filled with undefined values, and bitwise compare will fail.

I don't think we will use the bitwise comparison today. This path exist only because, otherwise, this code will fail to compile if we want to create an opaque object a concrete object without `operator==`.

For the future, if we encounter a case where we know it will work (for example, it is a packed POD), we can omit the `operator==`.

Reply:
@zasdfgbnm Can you please add this as a code comment?

Reply:
added